---
title: "User Guide"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{User Guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup}
library(cub)
```

# Input data

Prepare input genomes for *P. aeruginosa* *PAO1* and *PA14* strains, *E. coli* *K12 MG1655*:

```{r}
pao1 <- pao1
pa14 <- pa14
ek12 <- k12
```

Convert them to the `CubSet` objects:

```{r eval = FALSE}
cub01 <- CubSet(pao1)
cub14 <- CubSet(pa14)
cub12 <- CubSet(ek12)
```

Get the reference set of highly expressed genes (based on the work of Sharp and Li (1986) done on *E. coli*):

```{r eval = FALSE}
ref00 <- Sharp166$gene[Sharp166$expression == 'very high']
```

Find subset of such genes in all genomes, and calculate Relative Synonymous Codon Usage (RSCU) for each reference set. Compare it to the RSCU for randomly generated DNA:

```{r message = FALSE, eval = F}
library(TxDb.Paeruginosa.PAO1)
library(TxDb.Paeruginosa.PA14)
library(TxDb.Ecoli.K12.MG1655)
library(ggplot2)
library(ggpubr)

map01 <- biomaRt::select(TxDb.Paeruginosa.PAO1, names(pao1), 'TXNAME', 'CDSNAME')
map14 <- biomaRt::select(TxDb.Paeruginosa.PA14, names(pa14), 'TXNAME', 'CDSNAME')
map12 <- biomaRt::select(TxDb.Ecoli.K12.MG1655, names(ek12), 'TXNAME', 'CDSNAME')

idx01 <- as.numeric(na.omit(match(ref00, map01[, 2])))
idx14 <- as.numeric(na.omit(match(ref00, map14[, 2])))
idx12 <- as.numeric(na.omit(match(ref00, map12[, 2])))

ref01 <- CubSet(pao1[idx01]) |> rscu()
ref14 <- CubSet(pa14[idx14]) |> rscu()
ref12 <- CubSet(ek12[idx12]) |> rscu()

dna   <- sapply(ek12[idx12], length) 
dna   <- Biostrings::DNAStringSet(sapply(dna, \(x) { dnar(x) }))
refrd <- CubSet(dna) |> rscu()

prd <- cplot(refrd) + ggtitle('Randomly generated DNA') 
p01 <- cplot(ref01) + ggtitle('Reference set, P. aeruginosa PAO1')
p14 <- cplot(ref14) + ggtitle('Reference set, P. aeruginosa PA14')
p12 <- cplot(ref12) + ggtitle('Reference set, E. coli, K12 MG1655')

ggarrange(p01, p12, p14, prd, ncol = 2, nrow = 2) + ggtitle('Relative Synonymous Codon Usage (RSCU)')
```

### Tabular view for PAO1, PA14

Tabulate RACU for *P. aeruginosa PAO1 & PA14*.

```{r}
res <- dplyr::bind_rows(
  as.data.frame(ref01) |>
    dplyr::mutate(org = 'P. aeruginosa, PAO1'),
  as.data.frame(ref14) |>
    dplyr::mutate(org = 'P. aeruginosa, PA14')
  ) |>
  dplyr::filter(aa != "*") |>
  dplyr::mutate(aa = purrr::map_chr(aa, ~AMINOACIDS[[.x]][[2]])) |>
  dplyr::arrange(aa, dplyr::desc(val)) |>
  dplyr::mutate(aa = forcats::fct_reorder(aa, val, sum)) |>
  tidyr::pivot_wider(names_from = org, values_from = val) 

 plt <- ggplot2::ggplot(
    data = tidyr::pivot_longer(res, 3:4) |> dplyr::mutate(name = factor(name, levels = c('P. aeruginosa, PAO1', 'P. aeruginosa, PA14'))),
    mapping = ggplot2::aes(x = aa, y = value, label = codon)
  ) +
    ggplot2::geom_bar(stat = 'identity', position = 'stack', fill = 'gray35', color = 'gray85') +
    ggplot2::geom_text(
      stat = 'identity',
      position = ggplot2::position_stack(vjust = .5),
      size = 2.5,
      color = 'gray75'
    ) +
    ggplot2::scale_fill_grey() +
    ggplot2::facet_wrap(~name) +
    ggplot2::coord_flip() +
    ggplot2::labs(x = 'Amino acid', y = 'RSCU') +
    ggthemes::theme_few() +
    ggplot2::theme(axis.title.y = element_blank())
 
colnames(res) <- stringr::str_to_sentence(colnames(res))
xlsx::write.xlsx(as.data.frame(res), file = 'rscu.xlsx', sheetName = 'RSCU', row.names = FALSE)
```


## Calculate CAI

```{r eval = F}
res01 <- cai(cub01, ref01)
res14 <- cai(cub14, ref14)
res12 <- cai(cub12, ref12)

dt <- data.frame(
  cai = c(res01, res14, res12),
  org = c(
    rep('P. aeruginosa, PAO1', length(res01)),
    rep('P. aeruginosa, PA14', length(res14)),
    rep('E. coli, K12 MG1655', length(res12))
    )
  ) |>
  dplyr::filter(!is.na(cai)) 

pd <- ggplot(data = dt, aes(x = cai, fill = org, colour = org)) +
  geom_density(linewidth = .25, alpha = .5) +
  labs(title = 'Codone Adaptation Index', subtitle = 'Calcuated per each individual gene', x = 'CAI', y = 'Desinty') +
  ggthemes::theme_few()
pd
```


## Highly expressed genes

### P. aeruginosa PAO1

```{r, eval = F}
idx_high <- which(res01 >= quantile(res01, .95, na.rm = TRUE))
idx_low <- which(res01 <= quantile(res01, .05, na.rm = TRUE))
pao1_high <- biomaRt::select(TxDb.Paeruginosa.PAO1, names(pao1)[idx_high], 'TXNAME', 'CDSNAME')[, 2]
pao1_low <- biomaRt::select(TxDb.Paeruginosa.PAO1, names(pao1)[idx_low], 'TXNAME', 'CDSNAME')[, 2]

pth  <- KEGGREST::keggLink('pathway', 'pae')
gns  <- KEGGREST::keggList(paste('pae', pao1_high, sep = ':'))
dt   <- data.frame(path = pth[names(gns)], genes = names(gns), func = gns)
table(dt$func)
```

### P. aeruginosa PA14

```{r eval = F}
idx_high  <- which(res14 >= quantile(res14, .95, na.rm = TRUE))
idx_low   <- which(res14 <= quantile(res14, .05, na.rm = TRUE))
pa14_high <- biomaRt::select(TxDb.Paeruginosa.PA14, names(pa14)[idx_high], 'TXNAME', 'CDSNAME')[, 2]
pa14_low  <- biomaRt::select(TxDb.Paeruginosa.PA14, names(pa14)[idx_low], 'TXNAME', 'CDSNAME')[, 2]

pth  <- KEGGREST::keggLink('pathway', 'pau')
gns  <- KEGGREST::keggList(paste('pau', na.omit(pa14_high[1:250]), sep = ':'))
dt   <- data.frame(path = pth[names(gns)], genes = names(gns), func = gns)

table(dt$func)
```

### E. coli K12 MG1655

```{r eval = F}
idx_high  <- which(res12 >= quantile(res12, .95, na.rm = TRUE))
idx_low   <- which(res12 <= quantile(res12, .05, na.rm = TRUE))
pa14_high <- biomaRt::select(TxDb.Ecoli.K12.MG1655, names(ek12)[idx_high], 'TXNAME', 'CDSNAME')[, 2]
pa14_low  <- biomaRt::select(TxDb.Ecoli.K12.MG1655, names(ek12)[idx_low], 'TXNAME', 'CDSNAME')[, 2]
